pkg/xds/server/callbacks/dataplane_sync_tracker.go

```diff
@@ -30,35 +30,48 @@ var _ DataplaneCallbacks = &dataplaneSyncTracker{}
 // Node info can be (but does not have to be) carried only on the first XDS request. That's why need streamsAssociation map
 // that indicates that the stream was already associated
 type dataplaneSyncTracker struct {
-	NoopDataplaneCallbacks
-
 	newDataplaneWatchdog NewDataplaneWatchdogFunc
 
 	stdsync.RWMutex // protects access to the fields below
-	watchdogs       map[core_model.ResourceKey]context.CancelFunc
+	watchdogs       map[core_model.ResourceKey]*watchdogState
+}
+type watchdogState struct {
+	cancelFunc context.CancelFunc
+	stopped    chan struct{}
 }
 
+//nolint:contextcheck // it's not clear how the parent go-control-plane context lives
 func (t *dataplaneSyncTracker) OnProxyConnected(streamID core_xds.StreamID, dpKey core_model.ResourceKey, _ context.Context, _ core_xds.DataplaneMetadata) error {
 	// We use OnProxyConnected because there should be only one watchdog for given dataplane.
 	t.Lock()
 	defer t.Unlock()
 
 	ctx, cancel := context.WithCancel(context.Background())
-	t.watchdogs[dpKey] = func() {
-		dataplaneSyncTrackerLog.V(1).Info("stopping Watchdog for a Dataplane", "dpKey", dpKey, "streamID", streamID)
-		cancel()
+	state := &watchdogState{
+		cancelFunc: cancel,
+		stopped:    make(chan struct{}),
 	}
 	dataplaneSyncTrackerLog.V(1).Info("starting Watchdog for a Dataplane", "dpKey", dpKey, "streamID", streamID)
-	//nolint:contextcheck // it's not clear how the parent go-control-plane context lives
-	go t.newDataplaneWatchdog(dpKey).Start(ctx)
+	stoppedDone := state.stopped
+	go func() {
+		defer close(stoppedDone)
+		t.newDataplaneWatchdog(dpKey).Start(ctx)
+	}()
+	t.watchdogs[dpKey] = state
 	return nil
 }
 
-func (t *dataplaneSyncTracker) OnProxyDisconnected(_ context.Context, _ core_xds.StreamID, dpKey core_model.ResourceKey) {
-	t.Lock()
-	defer t.Unlock()
-	if cancelFn := t.watchdogs[dpKey]; cancelFn != nil {
-		cancelFn()
+func (t *dataplaneSyncTracker) OnProxyDisconnected(_ context.Context, streamID core_xds.StreamID, dpKey core_model.ResourceKey) {
+	t.RLock()
+	dpData := t.watchdogs[dpKey]
+	t.RUnlock()
+
+	if dpData != nil {
+		dpData.cancelFunc()
+		<-dpData.stopped
+		dataplaneSyncTrackerLog.V(1).Info("watchdog for a Dataplane stopped", "dpKey", dpKey, "streamID", streamID)
+		t.Lock()
+		defer t.Unlock()
+		delete(t.watchdogs, dpKey)
```

Comment:
It seems like there would be one concurrent problem:
```text
OnProxyDisconnected RLock --> OnProxyConnected Lock to set 't.watchdogs[dpKey]' --> OnProxyDisconnected Lock to delete 't.watchdogs'
```

Then the `t.watchdogs[dpKey]` would be deleted and it's in `Connected` state.

Please correct me if I'm wrong

Reply:
RLock will not impact Lock, and when it's Locked, a RLock will not be available. So when it's RLock-ed, we can still lock it in the OnProxyConnected event. The first part is correct.
But I don't think the second part will cause a problem since they are exclusive. 
The two Locking will wait for the lock to be released when it's locked by the other.

Reply:
OK, so if there is a second connection with the same DPKey established between after the RLock is released and the next Lock statement (for deleting the key), it can cleanup the wrong state.
This seems more like a concern that should be taken care of in the OnProxyConnected event

Reply:
OK, the upper dataplane_callback `activeStreams` will prevent this behavior, it's never go into `Connected` state